In [1]:
import numpy as np
import pandas as pd
from sklearn.metrics import accuracy_score  
from sklearn.metrics import recall_score  
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.linear_model import LogisticRegression 
from sklearn import neighbors
from sklearn import tree
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import BernoulliNB
import matplotlib.pyplot as plt

In [2]:
def get_the_predicted_bugs(name, num1, num2):
    All_Data = pd.read_csv(name, usecols = [i for i in range(num1, num2)]).values
    return All_Data

In [3]:
data = get_the_predicted_bugs('modified_all_data.csv', 2, 45)

## RandomForestClassifier

## best accuracy_score

In [4]:
data_random = np.random.permutation(data)
x = data_random[:, [i for i in range(data.shape[1] - 1)]]
y = data_random[:, -1]
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.4, random_state = 0)
rf_clf = RandomForestClassifier()

In [5]:
param_grid = [{'n_estimators': [1, 2, 3, 4, 5, 10, 200],
               'criterion': ['gini', 'entropy']}]

grid_search = GridSearchCV(rf_clf, param_grid, scoring = 'accuracy', n_jobs = -1, cv = 5)

In [6]:
grid_search.fit(x_train, y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False),
       fit_params=None, iid=True, n_jobs=-1,
       param_grid=[{'n_estimators': [1, 2, 3, 4, 5, 10, 200], 'criterion': ['gini', 'entropy']}],
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='accuracy', verbose=0)

In [7]:
grid_search.best_params_

{'criterion': 'entropy', 'n_estimators': 200}

In [8]:
y_pre = grid_search.best_estimator_.predict(x_test)
accuracy_score(y_test, y_pre), recall_score(y_test, y_pre)

(0.8731218697829716, 0.2)

## get the average accuracy_score

In [15]:
accuracy_score_list = []
recall_score_list = []
for i in range(100):
    data_random = np.random.permutation(data)
    x = data_random[:, [i for i in range(data.shape[1] - 1)]]
    y = data_random[:, -1]
    x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.4, random_state = 0)
    rf_clf = RandomForestClassifier(n_estimators = 200, criterion = 'entropy', n_jobs = -1)
    rf_clf.fit(x_train, y_train)
    
    y_pre = rf_clf.predict(x_test)
    accuracy_score_list.append(accuracy_score(y_test, y_pre))
    recall_score_list.append(recall_score(y_test, y_pre))

In [16]:
np.mean(accuracy_score_list)

0.8686143572621038

In [17]:
np.mean(recall_score_list)

0.16274439251958708

By RandomForestClassifier, aftering tuning the parameters for getting the best accuracy_score, the average accuracy_score of 100 times is about is about 0.8686 and the average recall_score of 100 times is about 0.16.

## best recall_score¶

In [18]:
data_random = np.random.permutation(data)
x = data_random[:, [i for i in range(data.shape[1] - 1)]]
y = data_random[:, -1]
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.4, random_state = 0)
rf_clf = RandomForestClassifier()

In [19]:
param_grid = [{'n_estimators': [1, 2, 3, 4, 5, 10, 200],
               'criterion': ['gini', 'entropy']}]

grid_search = GridSearchCV(rf_clf, param_grid, scoring = 'recall', n_jobs = -1, cv = 5)

In [20]:
grid_search.fit(x_train, y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False),
       fit_params=None, iid=True, n_jobs=-1,
       param_grid=[{'n_estimators': [1, 2, 3, 4, 5, 10, 200], 'criterion': ['gini', 'entropy']}],
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='recall', verbose=0)

In [21]:
grid_search.best_params_

{'criterion': 'entropy', 'n_estimators': 1}

In [22]:
y_pre = grid_search.best_estimator_.predict(x_test)
accuracy_score(y_test, y_pre), recall_score(y_test, y_pre)

(0.8046744574290484, 0.2727272727272727)

## get the average accuracy_score¶

In [23]:
data_random = np.random.permutation(data)
x = data_random[:, [i for i in range(data.shape[1] - 1)]]
y = data_random[:, -1]
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.4, random_state = 0)
rf_clf = RandomForestClassifier()

In [24]:
accuracy_score_list = []
recall_score_list = []
for i in range(100):
    data_random = np.random.permutation(data)
    x = data_random[:, [i for i in range(data.shape[1] - 1)]]
    y = data_random[:, -1]
    x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.4, random_state = 0)
    rf_clf = RandomForestClassifier(n_estimators = 1, criterion = 'entropy', n_jobs = -1)
    rf_clf.fit(x_train, y_train)
    
    y_pre = rf_clf.predict(x_test)
    accuracy_score_list.append(accuracy_score(y_test, y_pre))
    recall_score_list.append(recall_score(y_test, y_pre))

In [25]:
np.mean(accuracy_score_list)

0.8009348914858098

In [26]:
np.mean(recall_score_list)

0.2893665611581639

By RandomForestClassifier, aftering tuning the parameters for getting the best recall_score, the average accuracy_score of 100 times is about is about 0.80 and the average recall_score of 100 times is about 0.29.